flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -69,36 +68,34 @@
     NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connectionId)
             throws IOException, InterruptedException {
         while (true) {
-            AtomicBoolean isTheFirstOne = new AtomicBoolean(false);
-            CompletableFuture<NettyPartitionRequestClient> clientFuture =
-                    clients.computeIfAbsent(
-                            connectionId,
-                            unused -> {
-                                isTheFirstOne.set(true);
-                                return new CompletableFuture<>();
-                            });
-            if (isTheFirstOne.get()) {
+            final CompletableFuture<NettyPartitionRequestClient> newClientFuture =
+                    new CompletableFuture<>();
+
+            final CompletableFuture<NettyPartitionRequestClient> clientFuture =
+                    clients.putIfAbsent(connectionId, newClientFuture);
+
+            final NettyPartitionRequestClient client;
+
+            if (clientFuture == null) {
                 try {
-                    clientFuture.complete(connectWithRetries(connectionId));
-                } catch (InterruptedException e) {
-                    clientFuture.complete(null); // let others waiting know that they should retry
+                    client = connectWithRetries(connectionId);
+                } catch (RemoteTransportException | InterruptedException e) {
```

Comment:
My assumption was that uncaught exceptions will trigger a fatal failure. If this isn't the case, then we should indeed catch all exceptions here.

Reply:
I think it will fail the current job but not necessarily subsequent jobs using the same `ConnectionId`.
In either case, I would not bake in this assumption, as it can change and makes more difficult to reason about.

Reply:
I think you are right. I will update the PR.

Reply:
Thanks!

I just spotted another related issue: this exception is re-thrown below without being wrapped (at line 86).

Previously, it was fine because it was only `InterruptedException` (other exceptions were wrapped after `future.get`).
Not wrapping it into `IOException` can break the logic in `RestartPipelinedRegionFailoverStrategy.getTasksNeedingRestart`.
WDYT?

Reply:
I think this is fine because `RemoteTransportException` is an `IOException`. Any other exception which bubbles up from here will be unchecked and should lead to a fatal exception (at least on the Task level).

Reply:
I think you're right. Though other clients (threads) waiting for this future do `rethrowIOException`.

Reply:
`rethrowIOException` only rethrows checked exceptions as `IOExceptions`. All unchecked exceptions will remain unchanged. Hence, this shouldn't be a problem as well.